File: 297-serialize_and_deserialize_binary_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:
    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        def rser(root, serialized):
            if not root:
                serialized += "None,"
                return serialized

            serialized += str(root.val) + ","
            serialized = rser(root.left, serialized)
            serialized = rser(root.right, serialized)
            
            # print(serialized)
            return serialized

        return rser(root, "")
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        values_preorder = data.split(",")
        def rdeser(values_preorder):
            if values_preorder[0] == "None":
                values_preorder.pop(0)
                return None
            
            root = TreeNode(values_preorder.pop(0))
            root.left = rdeser(values_preorder)
            root.right = rdeser(values_preorder)

            return root

        return rdeser(values_preorder)
            

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: 297-serialize_and_deserialize_binary_tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public string serialize(TreeNode root)
    {
        StringBuilder rser(TreeNode node, StringBuilder s)
        {
            if (node == null) {
                s.Append("null,");
                return s;
            }

            s.Append(node.val + ",");
            s = rser(node.left, s) ;
            s = rser(node.right, s) ;

            return s;
        }

        var sb = rser(root, new());
        return sb.ToString();
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(string data)
    {
        TreeNode rdeser(Queue<string> nodes)
        {
            if (nodes.Count() == 0)
            {
                return null;
            }

            string val = nodes.Dequeue();
            if (val == "null")
            {
                return null;
            }
            
            TreeNode node = new(Convert.ToInt32(val));

            node.left = rdeser(nodes);
            node.right = rdeser(nodes);

            return node;
        }

        Queue<string> nodesPreorder = new(data.Split(","));

        return rdeser(nodesPreorder);
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));
